python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -613,6 +614,21 @@ def do_evaluate(
             return Column(
                 plc.strings.find.starts_with(column.obj, suffix.obj), column.name
             )
+        elif self.name == pl_expr.StringFunction.Contains:
+            col, substr = columns
+            literal, strict = self.options
+            if not strict:
+                # libcudf always errors if the pattern isn't valid
+                raise NotImplementedError("strict=False not supported in contains")
+            if literal:
+                return Column(plc.strings.find.contains(col.obj, substr.obj), col.name)
+            else:
+                # TODO: hack
+                pattern = plc.interop.to_arrow(substr.obj).as_py()
```

Comment:
@vyasr this suggests that `pylibcudf.strings.regex_program.RegexProgram.create` should accept a `pylibcudf.Scalar` rather than a `str`, what do you think?

Reply:
I think probably a fused type. Users should be able to provide a string for ease of use, but power users can provide a scalar to avoid the extra allocation.

Reply:
@wence- makes a good point. The libcudf api being targeted doesn't have an overload that accepts a device side `cudf::scalar`. I think this means we can't avoid moving to a host scalar to call the API at _some point_ in the process. 

As for how to do it, I'm perhaps weakly against adding more machinery to `plc.Scalar` itself and would be more in favor of adding some kind of private helper function perhaps in the `plc.scalar` module, like `_to_host_scalar(scalar: plc.Scalar) -> str, int, float...` I think that this might become a wider question since I believe there are other libcudf apis that accept host side scalars, certainly for strings but likely elsewhere.

Reply:
I think we want utilities that produce strongly typed scalars from the type-erased version (since there are a bunch of functions in libcudf that accept (say) a `string_scalar` rather than a `scalar`), the libcudf `string_scalar` then has a `string_view` (or similar) function that produces a host-side scalar, I think.

We could hide them behind `plc.Scalar.str` string methods namespace?

Reply:
Having thought about this for a few days I'm thinking there's a higher level question here around `Literal`s. Since most libcudf APIs accept a device scalar, it's worked so far for us to [eagerly copy a passed literal to device](https://github.com/rapidsai/cudf/blob/branch-24.08/python/cudf_polars/cudf_polars/dsl/expr.py#L366). But this one expects a host scalar, so regardless of _how_ we get the host scalar, we're still copying the scalar from host to device during `Literal.do_evaluate`, then from device _back_ to host to get the host value for the libcudf `contains_re` API (and presumably back to the device inside libcudf to actually call the API). 

@wence- Is my assessment of this correct? If so, I see during `do_evaluate`, we have a `context` object handy. Can we currently, or in some easy to reach state, use the context to distinguish these scenarios and avoid the extra two copies?

Reply:
Well, if you are raising in the constructor of `StringFunction` for the `contains_re` case when the pattern is not a `Literal`, then in the `evaluate` method, for `contains_re`, you can directly reach inside the children and pull out the (host-side) arrow scalar without evaluating the child. Something like:

```
# stringfunction
def do_evaluate(self, ...):
    if self is contains_re:
        child, pattern = self.children
        column = child.evaluate(...)
        assert isinstance(pattern, Literal)
        prog = RegexProgram(pattern.value)
        ...
```

Reply:
Done in https://github.com/rapidsai/cudf/pull/15918/commits/e3fb170598dbaaca32b2e43d482209416f94c965

Reply:
> , then from device _back_ to host to get the host value for the libcudf `contains_re` API (and presumably back to the device inside libcudf to actually call the API).

Just FYI. The regex pattern is never copied to the device inside libcudf. The pattern is _compiled_ into a set of states in CPU code that are then copied to the device. Since the device code never needs the original pattern, libcudf only accepts the pattern string in host memory.